### PR TITLE
BIP-352: handle invalid privkey / pubkey sums for sending / scanning, add changelog

### PR DESCRIPTION
**Pull request by theStack**
See the discussion in the corresponding secp256k1 PR (https://github.com/bitcoin-core/secp256k1/pull/1519#issuecomment-2142599641, https://github.com/bitcoin-core/secp256k1/pull/1519#issuecomment-2143167510, https://github.com/bitcoin-core/secp256k1/pull/1519#issuecomment-2144837459), where it was recommended to include this corner-case in the BIP by real-or-random:

> Indeed, this should be in the BIP, ideally even in the pseudocode. If our code starts to reject "payments", then better be authorized by the standard.

Note that this PR right now only includes the absolute minimum change; happy to include further changes if demanded (reflect this in the Python implementation, add test case), but not sure what are common rules for BIP amendments and if additional changes are worth it for a corner-case. For an additional test vector, one could take the example tx that has been published on signet: [d73f4a19f3973e90af6df62e735bb7b31f3d5ab8e7e26e7950651b436d093313](https://mempool.space/signet/tx/d73f4a19f3973e90af6df62e735bb7b31f3d5ab8e7e26e7950651b436d093313)

### CODE REVIEW

File: bip-0352.mediawiki [hunk 1]
```diff
@@ -337,6 +337,7 @@ If each of the checks in ''[[#scanning-silent-payment-eligible-transactions|Scan
 
 * Generate the ''input_hash'' with the smallest outpoint lexicographically, using the method described above
 * Let ''A = A<sub>1</sub> + A<sub>2</sub> + ... + A<sub>n</sub>'', where each ''A<sub>i</sub>'' is the public key of an input from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+** In case ''A'' is the point at infinity, the transaction is not silent payment eligible, and one should proceed with the next one.
```
Comment (josibake) [inline bip-0352.mediawiki:340]: From what I understand, the point at infinity is not a valid point on the curve, so it might be better to say something like:

```suggestion
** If ''A'' is not a valid point, i.e., if the sum of the input public keys results in the point at infinity, fail
```

**Review by josibake — COMMENTED** [review]
Thanks for pointing this out, @theStack ! If you are able to add a test case, that would be great! If not, no worries, I can add one later this week.

File: bip-0352.mediawiki (hunk 1 above)
Comment (real-or-random) [reply bip-0352.mediawiki:340]: Well, the point at infinity is certainly in the group (and thus a possible result of the addition). Whether it's "on the curve" depends on your definition of curve, and whether it's a "valid point" probably depends on the context. So I think the original text was slightly better (but I guess either is fine in the end).

**Comment by real-or-random** [comment]
> but not sure what are common rules for BIP amendments and if additional changes are worth it for a corner-case.

I'm not sure if there are common rules, but I think it's good practice to update BIPs to clarify such cases. And then if the text is changed, then the Python code should also be changed.

I don't want to make it too complicated, but it may be a good idea to include a changelog, like we did in BIP327: https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#change-log (this one looks long because we started versioning even before the BIP was merged into the repo here...)

File: bip-0352.mediawiki (hunk 1 above)
Comment (josibake) [reply bip-0352.mediawiki:340]: If "valid" is context specific, then I agree its best to mention the point at infinity specifically. "one should" feels a bit too much like a suggestion to me, perhaps something more like:

```suggestion
** If ''A'' is the point at infinity, skip the transaction
```

and then also add a bullet point to the section above, "Scanning silent payment eligible transactions"

**Comment by davidaucoin7377** [comment]
> See the discussion in the corresponding secp256k1 PR (https://github.com/bitcoin-core/secp256k1/pull/1519#issuecomment-2142599641, https://github.com/bitcoin-core/secp256k1/pull/1519#issuecomment-2143167510, https://github.com/bitcoin-core/secp256k1/pull/1519#issuecomment-2144837459), where it was recommended to include this corner-case in the BIP by real-or-random:
> 
> > Indeed, this should be in the BIP, ideally even in the pseudocode. If our code starts to reject "payments", then better be authorized by the standard.
> 
> 
> Note that this PR right now only includes the absolute minimum change; happy to include further changes if demanded (reflect this in the Python implementation, add test case), but not sure what are common rules for BIP amendments and if additional changes are worth it for a corner-case. For an additional test vector, one could take the example tx that has been published on signet: [d73f4a19f3973e90af6df62e735bb7b31f3d5ab8e7e26e7950651b436d093313](https://mempool.space/signet/tx/d73f4a19f3973e90af6df62e735bb7b31f3d5ab8e7e26e7950651b436d093313)

**Review by davidaucoin7377 — COMMENTED** [review]
Test run

File: bip-0352.mediawiki (hunk 1 above)
Comment (real-or-random) [reply bip-0352.mediawiki]: > and then also add a bullet point to the section above, "Scanning silent payment eligible transactions" 

nit: I don't think anyone will misinterpret this, but it's this is a tiny bit confusing because of the text just above the algorithm:
```
If each of the checks in ''[[#scanning-silent-payment-eligible-transactions|Scanning silent payment eligible transactions]]'' passes, the receiving wallet must:
```

It may be a cleaner not to add the bullet point. The checks in the bullet points seem to be rather syntactical checks, so I think it's okay to omit this one.

**Comment by real-or-random** [comment]
> and then also add a bullet point to the section above, "Scanning silent payment eligible transactions" 

nit: I don't think anyone will misinterpret this, but it's this is a tiny bit confusing because of the text just above the algorithm:
```
If each of the checks in ''[[#scanning-silent-payment-eligible-transactions|Scanning silent payment eligible transactions]]'' passes, the receiving wallet must:
```

It may be a cleaner not to add the bullet point. The checks in the bullet points seem to be rather syntactical checks, so I think it's okay to omit this one.


> I thought that it's maybe worth it to update the secp256k1 implementation to one that follows best practices and has a nicer API (I think you were working on such a Python module, that'd probably be a good choice?), but that's of course outside of the scope of this PR.

Yep, we'll hopefully have something soon because we also use it in our draft for the DKG BIP, which we want to post to the mailing list in the next weeks.

**Review by 5twelve — APPROVED** [review]

File: bip-0352/reference.py [hunk 2]
```diff
@@ -127,6 +127,10 @@ def create_outputs(input_priv_keys: List[Tuple[ECKey, bool]], input_hash: bytes,
         negated_keys.append(k)
 
     a_sum = sum(negated_keys)
+    if not a_sum.valid:
+        # Input privkeys sum is zero -> fail
+        return []
```
Comment (josibake) [inline bip-0352/reference.py:132]: Any reason we can't reuse the private keys corresponding to the public keys in the recipient test case here?

**Review by josibake — COMMENTED** [review]
Looks good! Since we are now checking for `a = 0`, I think its worth adding a test case for the sender side, as well.

Regarding the changelog, I think this is a great suggestion and especially helpful when adding new test cases. I'd suggest adding one here and just reusing the semantic versioning system from BIP327.

---
(I realize this is a bit out of scope for this change, but I didn't see a way to open an issue on the BIPs repo for this discussion)

More generally, seems worth it to formalizing the changelog format and versioning scheme as part of BIP2 (or whichever process BIP makes sense), e.g. something like "BIP should include a change log. If a change log is added it must use `MAJOR.MINOR.PATH` versioning scheme, where `MAJOR` is.. " etc.

I see a few BIPs have change logs in various formats, which got me thinking it would be nice to formalize it for consistency and also to avoid each BIP needing to either a) copy the text from another BIP explaining the change log or b) re-invent their own versioning scheme each time.

**Comment by josibake** [comment]
> Simple suggestion

LGTM! I don't think the pre-merge history is super relevant/useful here. Thanks for picking this up!

File: bip-0352.mediawiki [hunk 3]
```diff
@@ -483,6 +483,17 @@ A malicious notification could potentially cause the following issues:
...
 Wallet designers can choose which tradeoffs they find appropriate. For example, a wallet could check the block filter to at least probabilistically confirm the likely existence of the UTXO, thus efficiently cutting down on spam. The payment could then be marked as unconfirmed until a scan is performed and the existence of the UTXO in accordance to the silent payment specification is verified.
 
+== Change Log ==
+
+To help implementers understand updates to this document, we attach a version number that resembles ''semantic versioning'' (<code>MAJOR.MINOR.PATCH</code>).
+The <code>MAJOR</code> version is incremented if changes to the BIP are introduced that are incompatible with prior versions.
+The <code>MINOR</code> version is incremented whenever the inputs or the output of an algorithm changes in a backward-compatible way or new backward-compatible functionality is added.
+The <code>PATCH</code> version is incremented for other changes that are noteworthy (bug fixes, test vectors, important clarifications, etc.).
+
+* '''1.0.0''' (2024-05-08):
```
Comment (josibake) [inline bip-0352.mediawiki:493]: nit: In my experience, change logs usually start with the latest version first (which is also how the BIP327 change log is structured). Aside from consistency, I also prefer the latest first because as implementors most people are interested in the most recent changes and not necessarily interest in the chronology of changes

**Review by josibake — APPROVED** [review]
ACK https://github.com/bitcoin/bips/pull/1620/commits/4c0c3d7d77113507e32187dac662023e85f9c0d8

Test cases look good, left a nit regarding change log ordering, not blocking.
